ystia/yorc-a4c-plugin
# Bugfix/brbdcf 1129 yorc openstack fip modifier changes the workflow ordering

# Pull Request description

## Description of the change

Fixing an issue where the Yorc plugin when applying a TopologyModifier to replace the public network with a Floating IP, was breaking the dependency between a Compute and its attached Network,
causing wrong workflows where Compute and network install steps appear being run in parallel while the network install step is expected to run before the Compute install step.

Added info logs about the Topology Modification (adding Floating IP nodes), that will appear in Alien4Cloud when the user chooses the OpenStack Location.
For example, when deploying a Consul on Block Storage sample application described below, the following logs will appear in Alien4Cloud :

```
Automatic matching for template <ComputeCSDC1>
Automatic matching for template <BlockStorage>
Automatic matching for template <PublicNetwork>
Automatic matching for template <ComputeCADC1>
<FIPComputeCADC1> created to provide a Floating IP address to <ComputeCADC1> on network <PublicNetwork>
<FIPComputeCSDC1> created to provide a Floating IP address to <ComputeCSDC1> on network <PublicNetwork>
Found a BlockStorage <BlockStorage> with an attachment on <ComputeCSDC1>. Let's swap their workflow steps to match Yorc expectations. 
```

The two lines on Floating IP address described above are new, added by this Pull Request.

### What I did

Initially the code was creating a new Floating IP Node Template, and was re-using after modification the relationship between the Compute and the Network.
Then the code was deleting the Network Node Template.
This deletion had the side effect of deleting all relationships associated to the Network Node.
And so the Network-Compute relationship re-used to represent the FloatingIP-Compute was also deleted.

Instead of re-using the Network-Compute relationship, the code is now creating a new relationship FloatingIP-Compute.

### How to verify it

On a setup where the Yorc plugin was added to Alien4Cloud, Yorc was configured as an orchestrator, and an OpenStack location was configured.
From Alien4Cloud, Menu `Administration` > `Orchestrators`, select `Yorc`.
Then in the left-hand-side menu, select `Location`, `On demand resources`, and add the following resources :
  * yorc.nodes.openstack.PublicNetwork, provide a value for floating_network_name 
  * yorc.nodes.openstack.Compute, specify image, flavor, key_pair, credentials
  * yorc.nodes.openstack.BlockStorage, specify a size of 1 GiB.

From Alien4Cloud, Menu `Catalog` > `Manage Archives` > `Git import`
add the following Git location:
  * Repository URL: https://github.com/ystia/yorc-a4c-plugin/
  * Branch:  develop
  * Archive Folder: tosca-samples/org/ystia/yorc/samples

From the Menu `Applications`
Create a new Application, Initializing the topology from Topology Template `ConsulOnBS`.
Select the Environment, select the Location OpenStack.

Check the following Information messages appear in Alien4Cloud :
```
Automatic matching for template <ComputeCSDC1>
Automatic matching for template <BlockStorage>
Automatic matching for template <PublicNetwork>
Automatic matching for template <ComputeCADC1>
<FIPComputeCADC1> created to provide a Floating IP address to <ComputeCADC1> on network <PublicNetwork>
<FIPComputeCSDC1> created to provide a Floating IP address to <ComputeCSDC1> on network <PublicNetwork>
Found a BlockStorage <BlockStorage> with an attachment on <ComputeCSDC1>. Let's swap their workflow steps to match Yorc expectations. 
```

Click on `Review and Deploy`
Click on `Deploy`

Once the Deployment has started, Seletc on the left hand side Worflow
Check  the Step `ComputeCSDC1_Install` will happen only after steps `FIPComputeCSDC1_install` and `BlockStorage_install` completion.
Check the step `ComputeCADC1_Install` will happen only after step `FIPComputeCADC1_install` completion.

On the rigth-hand side, swith to Worflow `uninstall`, and check the dependencies are in reverse order :  BlockStorage and FIPCompute uninstall steps happen after Compute uninstall steps.

Wait for the deployment end, check it was successful.
